### PR TITLE
Feat/nico/sync contract

### LINKED ISSUE #30: Sync backend - contract
- [x] Sync the latest contract version with the backend
- [x] Try to automate the artifacts and abi.json retrieval from contracts repo?

### PR DESCRIPTION
**Pull request by NicoSerranoP**
1. Use `typechain` for contract types
2. Sync backend data to contract data
3. set `syncToBlockchain` in every interaction

**Note:** There are a few of pending items:
1. The likes `syncToBlockchain` is still pending, I haven't go deep into the implementation of nonces.
2. The EIP signature is not sent by the backend yet

Closes #30

### CODE REVIEW

File: src/controllers/contract.ts [hunk 1]
```diff
@@ -38,19 +38,17 @@ const syncDataToSmartContract = async (_req: Request, res: Response) => {
...
   } catch (error: any) {
     console.error("error: ", error);
     return res.status(500).json({ error: `Failed to sync state: ${error.message}` });
   }
 
   // Mark all as synced
-  await UserControl.markSynced(users);
-  await TagControl.markSynced(tags);
-  await URLControl.markSynced(urls);
+  await UserControl.markSynced(users.map(user => user.userAddress as string));
```
Comment (hackertron) [inline src/controllers/contract.ts:49]: Nico can you explain why this change? instead of complete users , tag and url object we are taking in only name, url and address? On smart contract side I assume only these 3 are needed right?

**Review by hackertron — APPROVED** [review]
LGTM , just left a comment .

File: src/controllers/contract.ts (hunk 1 above)
Comment (NicoSerranoP) [reply src/controllers/contract.ts:49]: The `getUsersToSync` is returning the format/type that the smart contract accepts. We get this type from the typechain generated (you need to run `npm run typechain` to run it. I added the instruction in the readme).

I decided that `markSync` functions should not use the typechain to avoid using a smart contract type in a function that does not interact with the contract. That is why the `markSync` functions usually use a string id rather than the object itself (user address, url, tag name, etc)
